app/jobs/results_recorder_job.rb

```diff
@@ -1,28 +1,45 @@
 # Responsibilities:
 # Update DB per event info.
-# Interpret replication logic.
 # Is this event the last needed for the DV to be complete?
 # If NO, do nothing further.
 # If YES, send a message to a non-job pub/sub queue.
 class ResultsRecorderJob < ApplicationJob
   queue_as :endpoint_events
+  attr_accessor :pc, :pcs
+
+  before_perform do |job|
+    job.pcs ||= PreservedCopy
+                .by_druid(job.arguments.first)
+                .joins(endpoint: [:endpoint_type])
+                .where(
+                  'endpoint_types.endpoint_class' => 'archive',
+                  'preserved_copies.version' => job.arguments.second
+                )
+    job.pc ||= pcs.find_by!('endpoints.delivery_class' => Object.const_get(job.arguments.third))
+  end
 
   # @param [String] druid
   # @param [Integer] version
-  # @param [String] endpoint
+  # @param [String] delivery_class Name of the worker class that performed delivery
   # @param [String] checksum
-  def perform(druid, version, endpoint, checksum)
-    hash = {
+  def perform(druid, version, _delivery_class, _checksum)
+    raise "Status shifted underneath replication: #{pc.inspect}" unless pc.unreplicated?
+    pc.ok!
```

Comment:
d'oh, yes, good point.

not to get down too much of a rabbit-hole, but would an optimistic lock prevent other things from locking the row, while also letting other things read it?  i.e., another process could query that row in the DB, but this same process (or another audit process) wouldn't be able to obtain a lock of its own in preparation for the updates it might do?  i'm not particularly well-versed in RDBMS locking semantics.

Reply:
Right, an optimistic lock is the kind that just makes sure that a fingerprint (timestamp/increment/whatever) on the in-memory object matches the database row when it goes to update (otherwise fails).  Faked up example:
```SQL
UPDATE mytable 
SET field='value', fingerprint=24
WHERE (id=66, fingerprint=23)"
```
If the fingerprint has already been shifted the update cannot land.  When it does land, the fingerprint is shifted.

Optimistic is lighter-weight but less bulletproof.  You have to have some tolerance for eventual collision and failure.  Pessimistic locking uses the database for a mutex.  More costly, more rigorous, but when it collides you get deadlock.

Reply:
good to know, thanks!